examples/petstore-expanded/api/petstore-client.gen.go

```diff
@@ -108,6 +109,291 @@ func (c *Client) FindPetById(ctx context.Context, id int64) (*http.Response, err
+func (c *ClientWithResponses) AddPetWithResponse(ctx context.Context, body NewPet) (*AddPetResponse, error) {
+	rsp, err := c.AddPet(ctx, body)
+	if err != nil {
+		return nil, err
+	}
+	return ParseAddPetResponse(rsp)
+}
+
+// DeletePetResponse is returned by Client.DeletePet()
+type DeletePetResponse struct {
+	Body         []byte
+	HTTPResponse *http.Response
+	JSONDefault  Error // 'unexpected error' (#/components/schemas/Error)
+}
+
+// Status returns HTTPResponse.Status
+func (r *DeletePetResponse) Status() string {
+	if r.HTTPResponse != nil {
+		return r.HTTPResponse.Status
+	}
+	return http.StatusText(0)
+}
+
+// StatusCode returns HTTPResponse.StatusCode
+func (r *DeletePetResponse) StatusCode() int {
+	if r.HTTPResponse != nil {
+		return r.HTTPResponse.StatusCode
+	}
+	return 0
+}
+
+// ParseDeletePetResponse parses an HTTP response from a DeletePetWithResponse call
+func ParseDeletePetResponse(rsp *http.Response) (*DeletePetResponse, error) {
+	bodyBytes, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &DeletePetResponse{
+		Body:         bodyBytes,
+		HTTPResponse: rsp,
+	}
+
+	switch {
+	case rsp.StatusCode == 204:
+		break // No content-type
+	case strings.Contains(rsp.Header.Get("Content-Type"), "json"):
+		response.JSONDefault = Error{}
+		if err := json.Unmarshal(bodyBytes, &response.JSONDefault); err != nil {
+			return nil, err
+		}
+	}
+
+	return response, nil
+}
+
+// DeletePet request returning *DeletePetResponse
+func (c *ClientWithResponses) DeletePetWithResponse(ctx context.Context, id int64) (*DeletePetResponse, error) {
+	rsp, err := c.DeletePet(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	return ParseDeletePetResponse(rsp)
+}
+
+// FindPetByIdResponse is returned by Client.FindPetById()
+type FindPetByIdResponse struct {
+	Body         []byte
+	HTTPResponse *http.Response
+	JSON200      struct {
+		// Embedded struct due to allOf(#/components/schemas/NewPet)
+		NewPet
+		// Embedded fields due to inline allOf schema
+		Id int64 `json:"id"`
+	} // 'pet response' (#/components/schemas/Pet)
+	JSONDefault Error // 'unexpected error' (#/components/schemas/Error)
+}
+
+// Status returns HTTPResponse.Status
+func (r *FindPetByIdResponse) Status() string {
+	if r.HTTPResponse != nil {
+		return r.HTTPResponse.Status
+	}
+	return http.StatusText(0)
+}
+
+// StatusCode returns HTTPResponse.StatusCode
+func (r *FindPetByIdResponse) StatusCode() int {
+	if r.HTTPResponse != nil {
+		return r.HTTPResponse.StatusCode
+	}
+	return 0
+}
+
+// ParseFindPetByIdResponse parses an HTTP response from a FindPetByIdWithResponse call
+func ParseFindPetByIdResponse(rsp *http.Response) (*FindPetByIdResponse, error) {
+	bodyBytes, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	response := &FindPetByIdResponse{
+		Body:         bodyBytes,
+		HTTPResponse: rsp,
+	}
+
+	switch {
+	case strings.Contains(rsp.Header.Get("Content-Type"), "json") && rsp.StatusCode == 200:
```

Comment:
I think we need one more little tweak in the generated code.

When you get a response, eg, `FindPetByIdResponse`, and it has multiple possible response types, you have to check them against the zerovalue to know if they're set, or you look into the response bytes themselves to figure out which it is. If we change this in the future, it will be a breaking API change, so we want to take care of this now.

How about all fields on the Response objects being pointer types? This way, you can case off which isn't nil to figure out which response type you got.

Reply:
Done